src/mapreduce.jl

```diff
@@ -159,27 +159,46 @@ end
 (-)(A::DArray, x::Number) = A .- x
 (-)(x::Number, A::DArray) = x .- A
 
-map_localparts(f::Callable, d::DArray) = DArray(i->f(localpart(d)), d)
-map_localparts(f::Callable, d1::DArray, d2::DArray) = DArray(d1) do I
-    f(localpart(d1), localpart(d2))
+function map_localparts(f::Callable, d::DArray)
+    r = Base.asyncmap!(similar(procs(d), Future), procs(d)) do p
+        remotecall_wait(p) do
+            f(localpart(d))
+        end
+    end
+    return DArray(r)
```

Comment:
This will slow down the case where the mapping retains the size. We now have two network hops to each node for all situations which is not really required for the size retaining case.

Maybe introduce a keyword argument to `DArray(init, d::DArray; similar=true)`? In the default case, as now, the assumption is that the shape and size are similar. If `similar=false` we can recompute from the refs.